src/main/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperations.scala

```diff
@@ -156,64 +143,94 @@ object TezosDatabaseOperations {
+  
   /**
     * Given the operation kind, return range of fees and timestamp for that operation.
     * @param kind  Operation kind
     * @return      The average fees for a given operation kind, if it exists
     */
-  def calculateAverageFees(kind: String): Option[AverageFees] = {
-    val operationKinds = Set[String]{kind}
-    val action = for {
-      o <- Tables.Operations
-      if o.kind.inSet(operationKinds)
-    } yield (o.fee, o.timestamp)
-    val op = dbHandle.run(action.distinct.sortBy(_._2.desc).take(numberOfFeesAveraged).result)
-    val results = Await.result(op, Duration.apply(awaitTimeInSeconds, SECONDS))
-    val resultNumbers = results.map(x => x._1.getOrElse("0").toDouble)
-    val m: Int = ceil(mean(resultNumbers)).toInt
-    val s: Int = ceil(stdev(resultNumbers)).toInt
-    results.length match {
-      case 0 => None
-      case _ =>
-        val timestamp = results.head._2
-        Some(AverageFees(max(m - s, 0), m, m + s, timestamp, kind))
+  def calculateAverageFeesIO(kind: String)(implicit ec: ExecutionContext): DBIO[Option[AverageFees]] = {
+    def computeAverage(ts: java.sql.Timestamp, fees: Seq[(Option[String], java.sql.Timestamp)]): AverageFees = {
+      val values = fees.map {
+        case (fee, _) => fee.map(_.toDouble).getOrElse(0.0)
+      }
+      val m: Int = ceil(mean(values)).toInt
+      val s: Int = ceil(stdev(values)).toInt
+      AverageFees(max(m - s, 0), m, m + s, ts, kind)
     }
-  }
 
+    val opQuery =
+      Tables.Operations
+        .filter(_.kind === kind)
+        .map(o => (o.fee, o.timestamp))
+        .distinct
+        .sortBy { case (_, ts) => ts.desc }
+        .take(numberOfFeesAveraged)
+        .result
+
+    opQuery.map { 
+      timestampedFees =>
+        timestampedFees.headOption.map {
+          case (_, latest) =>
+            computeAverage(latest, timestampedFees)
+        }
+    }
+  }
 
   /**
     * Delete all accounts in database not associated with block at maxLevel.
-    * @return No value, database update
+    * @return the number of rows removed
     */
-  def purgeOldAccounts(): Try[Int] = {
-    ApiOperations.fetchMaxBlockLevelForAccounts().flatMap{ blockLevel =>
-      Try {
-        val query = Tables.Accounts.filter(row => !(row.blockLevel === blockLevel))
-        val action = query.delete
-        val op = dbHandle.run(action)
-        Await.result(op, Duration.apply(awaitTimeInSeconds, SECONDS))
-      }
+  def purgeOldAccounts()(implicit ex: ExecutionContext): Future[Int] = {
+    val purged = dbHandle.run {
+      val maxLevel = SupportQueries.accountsMaxBlockLevel
+      Tables.Accounts.filter(_.blockLevel =!= maxLevel).delete
     }
+    purged.andThen {
+      case Success(howMany) => logger.info("{} accounts where purged from old block levels.", howMany)
+      case Failure(e) => logger.error("Could not purge old block-levels accounts", e)
+    }
+  }
+
+  /** convertions from domain objects to database row format */
+  object RowConversion {
```

Comment:
Should we do this sort of thing for all row conversions? What's the rationale here?

Reply:
as explained in previous comments, my idea is that we can expect such conversions between "domain types"  and "row types" to increase in time.

Keeping such code in a clear scope and separated from the rest makes everything neater, as I see it.

Please let me know what you think